lucene/core/src/java/org/apache/lucene/search/WANDScorer.java

```diff
@@ -257,13 +288,17 @@ public int advance(int target) throws IOException {
             // Advance 'head' as well
             advanceHead(target);
 
-            // Pop the new 'lead' from 'head'
-            moveToNextCandidate(target);
+            if (scoreMode == ScoreMode.TOP_SCORES) {
+              // Update score bounds if necessary so
+              updateMaxScoresIfNecessary(target);
 
-            if (doc == DocIdSetIterator.NO_MORE_DOCS) {
-              return DocIdSetIterator.NO_MORE_DOCS;
+              if (doc == DocIdSetIterator.NO_MORE_DOCS) {
```

Comment:
we may have a problem, or maybe this line is not needed anymore

Before, `moveToNextCandidate` would advance `doc`. But now that you are calling `updateMaxScoresIfNecessary ` instead, which doesn't advance doc, it shouldn't be possible that doc == NO_MORE_DOCS on this line?

Reply:
For this one I actually moved the doc advancement to `NO_MORE_DOCS` for condition `head.size() == 0` into `updateMaxScoresIfNecessary`, as initially I felt these two are closely related and only applicable for TOP_SCORES mode (I probably should have updated the method name though) https://github.com/zacharymorn/lucene-solr/blob/9243ef01de6f4a4154275919a32f5703e970f49d/lucene/core/src/java/org/apache/lucene/search/WANDScorer.java#L463-L468 . And thus the original `moveToNextCandidate` implementation was separated into `updateMaxScoresIfNecessary` and `setDocAndFreq`. 

However, in hindsight I see this is not a very good abstraction, and it's probably better to keep the original `moveToNextCandidate` and handle the non-scoring mode inside. Let me give that a try.

Reply:
I just pushed a commit to revert back to the original `moveToNextCandidate` abstraction, and handle the non TOP_SCORES mode inside that method. I think that really helps to localize / simplify the changes (and reduce 1 more branching) a lot!  Please let me know if this looks good to you.